File: app/src/jvx/java/art/coded/wireframe/model/ElementRepository.java
package art.coded.wireframe.model;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import java.util.List;

import art.coded.wireframe.model.entity.Element;
import art.coded.wireframe.model.local.ElementDao;
import art.coded.wireframe.model.local.ElementRoomDatabase;

public class ElementRepository {

    private static final String LOG_TAG = ElementRepository.class.getSimpleName();

    private ElementDao mElementDao;
    private LiveData<List<Element>> mAllElements;

    public ElementRepository(Application application) {
        ElementRoomDatabase db = ElementRoomDatabase.getInstance(application);
        mElementDao = db.elementDao();
        mAllElements = mElementDao.getAll();
    }

    public LiveData<List<Element>> getAllElements() { return mAllElements; }
    public void insert(Element element) { new InsertAsyncTask(mElementDao).execute(element); }
    public void delete(Element element) { new DeleteAsyncTask(mElementDao).execute(element); }
    public void deleteAll() { new DeleteAllAsyncTask(mElementDao).execute(); }
    public LiveData<PagedList<Element>> getPagedList() {
        return new LivePagedListBuilder<>(mElementDao.pagingSource(), 15).build();
    }

    private static class InsertAsyncTask extends AsyncTask<Element, Void, Void> {

        ElementDao mAsyncTaskDao;
        public InsertAsyncTask(ElementDao dao) { mAsyncTaskDao = dao; }
        @Override protected Void doInBackground(Element... elements) {
            mAsyncTaskDao.insert(elements[0]);
            return null;
        }
    }

    private static class DeleteAsyncTask extends AsyncTask<Element, Void, Void> {

        ElementDao mAsyncTaskDao;
        public DeleteAsyncTask(ElementDao dao) { mAsyncTaskDao = dao; }
        @Override protected Void doInBackground(Element... elements) {
            mAsyncTaskDao.delete(elements[0]);
            return null;
        }
    }

    private static class DeleteAllAsyncTask extends AsyncTask<Void, Void, Void> {

        ElementDao mAsyncTaskDao;
        public DeleteAllAsyncTask(ElementDao dao) { mAsyncTaskDao = dao; }
        @Override protected Void doInBackground(Void... voids) {
            mAsyncTaskDao.deleteAll();
            return null;
        }
    }
}

File: app/src/ktx/java/art/coded/wireframe/model/ElementRepository.kt
package art.coded.wireframe.model

import androidx.lifecycle.LiveData
import art.coded.wireframe.model.local.ElementDao
import androidx.paging.PagedList
import androidx.paging.LivePagedListBuilder
import art.coded.wireframe.model.entity.Element
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

private val LOG_TAG = ElementRepository::class.java.simpleName

class ElementRepository(private val elementDao: ElementDao) {

    val allElements: LiveData<List<Element>> = elementDao.all

    suspend fun insert(element: Element) { withContext(Dispatchers.IO) { elementDao.insert(element) } }
    suspend fun delete(element: Element) { withContext(Dispatchers.IO) { elementDao.delete(element) } }
    suspend fun deleteAll() { withContext(Dispatchers.IO) { elementDao.deleteAll() } }
    val pagedList: LiveData<PagedList<Element>>
        get() = LivePagedListBuilder(
            elementDao.pagingSource(), 15
        ).build()
}